### PR TITLE
also link to CUDA::cufft_static in case of BUILD_SHARED_LIBS=OFF

### PR DESCRIPTION
**Pull request by gregorburger**
In case BUILD_SHARED_LIBS is disabled, the cufft library is still linked to the dynamic version, whereas all other cuda libraries are linked to the static versions by appending CUDA_LIB_EXT. This PR makes cufft handles the same way as all other cuda libraries. 

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/cudaarithm/CMakeLists.txt [hunk 1]
```diff
@@ -18,7 +18,7 @@ if(ENABLE_CUDA_FIRST_CLASS_LANGUAGE)
   endif()
   if(HAVE_CUFFT)
     # static version requires seperable compilation which is incompatible with opencv's current library structure
```
[the remark this answers is not in the document]
Comment (gregorburger) [reply modules/cudaarithm/CMakeLists.txt:20]: Oh bummer, haven't seen that! But it builds for us. We use a heavily reduced build, but cudaarithm is included. 
tested using cuda toolkit 11.7.1 and cmake 3.29.2.

I will test with 12 as well, but is there a specific test I should run?

[the remark this answers is not in the document]
Comment (cudawarped) [reply modules/cudaarithm/CMakeLists.txt:20]: I can't remember where it failed.  I just remember having lots of issues when I tried to include the static FFT library on Linux.  I think it was OK on Windows.

[the remark this answers is not in the document]
Comment (gregorburger) [reply modules/cudaarithm/CMakeLists.txt:20]: I got the error now! I also may have a workaround that would make the cmake a little more ugly. Apparently, there is a cufft_static_nocallback.a library that can be linked with the host compiler [0]. It seems opencv does not use these callbacks and the tests opencv_test_cudaarithm build and run and don't seem to have a dynamic dependency on any cuda libraries. 

[0] https://docs.nvidia.com/cuda/cufft/index.html#static-library-without-callback-support

[the remark this answers is not in the document]
Comment (gregorburger) [reply modules/cudaarithm/CMakeLists.txt]: I updated the PR, what do you think?

**Comment by katharinaHochf** [comment]
In any case someone uses older cmake versions:
CUDA::cufft_static_nocallback starting in CUDA 9.2, requires CMake 3.23+

otherwise static linking does not work even with that fix.

File: modules/cudaarithm/CMakeLists.txt [hunk 2]
```diff
@@ -17,8 +17,13 @@ if(ENABLE_CUDA_FIRST_CLASS_LANGUAGE)
...
   endif()
   if(HAVE_CUFFT)
-    # static version requires seperable compilation which is incompatible with opencv's current library structure
-    list(APPEND optional_dependencies CUDA::cufft)
+    if(BUILD_SHARED_LIBS)
+      list(APPEND optional_dependencies CUDA::cufft)
+    else()
+      # static version requires seperable compilation which is incompatible with opencv's current library structure
+      # the cufft_static_nocallback variant does not requires seperable compilation. callbacks are currently not used.
+      list(APPEND optional_dependencies CUDA::cufft_static_nocallback)
```
[the remark this answers is not in the document]
Comment (gregorburger) [reply modules/cudaarithm/CMakeLists.txt]: - should work in windows
- should work when cmake is too old
- should work when cuda is too old

**Comment by gregorburger** [comment]
> In any case someone uses older cmake versions: CUDA::cufft_static_nocallback starting in CUDA 9.2, requires CMake 3.23+
> 
> otherwise static linking does not work even with that fix.
> 
> https://cmake.org/cmake/help/latest/module/FindCUDAToolkit.html#cuda-toolkit-cufft

Should work with the latest changes as well. If the proper versions of cuda or cmake aren't met, it will fall back to dynamic libraries as was the behaviour previously.

File: modules/cudaarithm/CMakeLists.txt [hunk 3]
```diff
@@ -9,6 +9,9 @@ ocv_warnings_disable(CMAKE_CXX_FLAGS /wd4127 /wd4324 /wd4512 -Wundef -Wmissing-d
 set(extra_dependencies "")
 set(optional_dependencies "")
 if(ENABLE_CUDA_FIRST_CLASS_LANGUAGE)
+  if(UNIX AND NOT BUILD_SHARED_LIBS AND CUDA_VERSION_STRING VERSION_GREATER_EQUAL 9.2 AND CMAKE_VERSION VERSION_GREATER_EQUAL 3.23)
```
Comment (cudawarped) [inline modules/cudaarithm/CMakeLists.txt:12]: `CUDA_VERSION_STRING` is set as part of [FindCUDA ](https://cmake.org/cmake/help/latest/module/FindCUDA.html#result-variables).  When CUDA is a first class language and you are calling [FindCUDAToolkit](https://cmake.org/cmake/help/latest/module/FindCUDAToolkit.html#findcudatoolkit) you need to check the [`CUDAToolkit_VERSION...`](https://cmake.org/cmake/help/latest/module/FindCUDAToolkit.html#result-variables)

Comment (gregorburger) [reply modules/cudaarithm/CMakeLists.txt:12]: it is set here https://github.com/opencv/opencv/blob/bef5a8768004c3d9b5eeb742446afe7d4e3e8145/cmake/OpenCVDetectCUDALanguage.cmake#L34
so it should work.

Comment (cudawarped) [reply modules/cudaarithm/CMakeLists.txt:12]: Fair enough that can be refactored out in the future when *FindCuda* is completely removed.

**Review by cudawarped — APPROVED** [review]
LGTM 👍

**Comment by cudawarped** [comment]
> @cudawarped I'm new to the OpenCV workflow, is there anything missing starting a build on the CI?

No.  One of the core maintainers needs to have time to check this is sensible before kicking off the CI.  As this is the contrib repo this in not as high a priority.

**Comment by asmorkalov** [comment]
@gregorburger Thanks a lot for the contribution! Could you push some change to the pr. Github Actions stuck and cannot be re-triggered.

**Comment by asmorkalov** [comment]
@gregorburger Please rebase the PR to current 4.x.

**Pull request merged**